File: sdk/python/failproofai_sdk/evaluator/authoring.py
"""Evaluator definition registry and typed author results."""

from __future__ import annotations

import hashlib
import inspect
import json
import math
import re
from collections.abc import Awaitable, Callable, Mapping
from dataclasses import dataclass, field
from typing import Any

from failproofai_sdk.evaluator.protocol import (
    MAX_CATALOG_DEFINITIONS,
    MAX_DESCRIPTION_BYTES,
    MAX_DISPLAY_NAME_BYTES,
    MAX_DISPLAY_VALUE_BYTES,
    MAX_EVAL_KEY_BYTES,
    MAX_LABEL_BYTES,
    MAX_LABELS_PER_RESULT,
    MAX_REASONING_BYTES,
    MAX_RESULTS_PER_RUN,
    MAX_SUMMARY_BYTES,
    MAX_UNIT_BYTES,
    MAX_VERSION_BYTES,
    CatalogDefinition,
    ResultItem,
    ResultKind,
    SessionTranscript,
)

_KEY = re.compile(r"^[a-z][a-z0-9_]*$")
EvalFunction = Callable[[SessionTranscript], "EvalResult | Awaitable[EvalResult]"]
ConditionFunction = Callable[
    [SessionTranscript], "bool | ConditionResult | Awaitable[bool | ConditionResult]"
]
CancellationFunction = Callable[[SessionTranscript], "Any | Awaitable[Any]"]


def _bounded(value: str, *, field_name: str, maximum: int) -> str:
    if not isinstance(value, str):
        raise TypeError(f"{field_name} must be a string")
    if not value:
        raise ValueError(f"{field_name} must not be empty")
    size = len(value.encode("utf-8"))
    if size > maximum:
        raise ValueError(f"{field_name} is {size} bytes; maximum is {maximum}")
    # Reject C0 control characters and DEL, matching the server's `check_bounded`
    # (server/src/evaluator/protocol.rs). Without this the SDK accepts a string —
    # e.g. reasoning/summary quoting transcript text that contains an ANSI escape
    # or NUL — that the server then rejects with a NON-RETRYABLE 422, so a
    # successful evaluation is silently lost and its assignment dead-letters.
    # TAB, LF and CR are kept because real multi-line reasoning uses them.
    bad = next(
        (
            ch
            for ch in value
            if (ord(ch) < 0x20 and ch not in "\t\n\r") or ord(ch) == 0x7F
        ),
        None,
    )
    if bad is not None:
        raise ValueError(
            f"{field_name} must not contain control characters "
            f"(found U+{ord(bad):04X})"
        )
    return value


def _finite(value: float, field_name: str) -> float:
    if isinstance(value, bool) or not isinstance(value, (int, float)):
        raise TypeError(f"{field_name} must be a number")
    result = float(value)
    if not math.isfinite(result):
        raise ValueError(f"{field_name} must be finite")
    return result


def _labels(values: tuple[str, ...] | list[str]) -> tuple[str, ...]:
    if len(values) > MAX_LABELS_PER_RESULT:
        raise ValueError(f"at most {MAX_LABELS_PER_RESULT} labels are allowed")
    normalized = []
    for label in values:
        normalized.append(_bounded(label, field_name="label", maximum=MAX_LABEL_BYTES))
    if len(set(normalized)) != len(normalized):
        raise ValueError("labels must be unique")
    return tuple(sorted(normalized))


@dataclass(frozen=True)
class Score:
    value: float
    passed: bool | None = None
    unit: str = "ratio"
    display_value: str | None = None
    description: str | None = None

    def __post_init__(self) -> None:
        value = _finite(self.value, "score value")
        if not 0 <= value <= 1:
            raise ValueError("score value must be between 0 and 1")
        object.__setattr__(self, "value", value)
        if self.passed is not None and not isinstance(self.passed, bool):
            raise TypeError("score passed must be a boolean or None")
        _validate_result_text(self.unit, self.display_value, self.description)


@dataclass(frozen=True)
class Metric:
    value: float
    unit: str = ""
    display_value: str | None = None
    description: str | None = None

    def __post_init__(self) -> None:
        object.__setattr__(self, "value", _finite(self.value, "metric value"))
        _validate_result_text(self.unit, self.display_value, self.description)


@dataclass(frozen=True)
class Assertion:
    passed: bool
    description: str | None = None

    def __post_init__(self) -> None:
        if not isinstance(self.passed, bool):
            raise TypeError("assertion passed must be a boolean")
        if self.description is not None:
            _bounded(
                self.description,
                field_name="description",
                maximum=MAX_DESCRIPTION_BYTES,
            )


@dataclass(frozen=True)
class ConditionResult:
    applicable: bool
    reason_code: str = "condition_false"

    def __post_init__(self) -> None:
        if not isinstance(self.applicable, bool):
            raise TypeError("condition applicable must be a boolean")
        _validate_key(self.reason_code, "condition reason code")


def _validate_result_text(
    unit: str, display_value: str | None, description: str | None
) -> None:
    if unit:
        _bounded(unit, field_name="unit", maximum=MAX_UNIT_BYTES)
    if display_value is not None:
        _bounded(
            display_value,
            field_name="display value",
            maximum=MAX_DISPLAY_VALUE_BYTES,
        )
    if description is not None:
        _bounded(
            description,
            field_name="description",
            maximum=MAX_DESCRIPTION_BYTES,
        )


@dataclass(frozen=True)
class EvalResult:
    score: Score | None = None
    metrics: Mapping[str, Metric | float] = field(default_factory=dict)
    assertions: Mapping[str, Assertion | bool] = field(default_factory=dict)
    reasoning: str | None = None
    summary: str | None = None
    labels: tuple[str, ...] = ()

    def __post_init__(self) -> None:
        if self.reasoning is not None:
            _bounded(
                self.reasoning,
                field_name="reasoning",
                maximum=MAX_REASONING_BYTES,
            )
        if self.summary is not None:
            _bounded(self.summary, field_name="summary", maximum=MAX_SUMMARY_BYTES)
        object.__setattr__(self, "labels", _labels(list(self.labels)))

    def result_items(self, eval_key: str) -> tuple[ResultItem, ...]:
        items: list[ResultItem] = []
        if self.score is not None:
            items.append(
                ResultItem(
                    result_key=eval_key,
                    result_kind=ResultKind.SCORE,
                    numeric_value=self.score.value,
                    bool_value=self.score.passed,
                    unit=self.score.unit,
                    display_value=self.score.display_value,
                    description=self.score.description,
                    reasoning=self.reasoning,
                    labels=self.labels,
                )
            )
        for key, raw_metric in sorted(self.metrics.items()):
            _validate_key(key, "metric key")
            metric = (
                raw_metric if isinstance(raw_metric, Metric) else Metric(raw_metric)
            )
            items.append(
                ResultItem(
                    result_key=key,
                    result_kind=ResultKind.METRIC,
                    numeric_value=metric.value,
                    unit=metric.unit,
                    display_value=metric.display_value,
                    description=metric.description,
                    # A metric-kind eval's primary result IS the metric whose
                    # key equals eval_key; attach the eval's reasoning there so
                    # it is not silently dropped for non-score evals.
                    reasoning=self.reasoning if key == eval_key else None,
                    labels=self.labels,
                )
            )
        for key, raw_assertion in sorted(self.assertions.items()):
            _validate_key(key, "assertion key")
            assertion = (
                raw_assertion
                if isinstance(raw_assertion, Assertion)
                else Assertion(raw_assertion)
            )
            items.append(
                ResultItem(
                    result_key=key,
                    result_kind=ResultKind.ASSERTION,
                    bool_value=assertion.passed,
                    description=assertion.description,
                    # An assertion-kind eval's primary result is the assertion
                    # whose key equals eval_key; carry the eval's reasoning there
                    # so a non-score eval does not lose it.
                    reasoning=self.reasoning if key == eval_key else None,
                    labels=self.labels,
                )
            )
        if not items:
            raise ValueError("an EvalResult must contain a score, metric, or assertion")
        if len(items) > MAX_RESULTS_PER_RUN:
            raise ValueError(
                f"an EvalResult may contain at most {MAX_RESULTS_PER_RUN} results"
            )
        keys = [item.result_key for item in items]
        if len(keys) != len(set(keys)):
            raise ValueError("result keys must be unique within one evaluation run")
        return tuple(items)


def _validate_key(value: str, field_name: str = "eval_key") -> str:
    _bounded(value, field_name=field_name, maximum=MAX_EVAL_KEY_BYTES)
    if not _KEY.fullmatch(value):
        raise ValueError(f"{field_name} must match {_KEY.pattern}")
    return value


@dataclass(frozen=True)
class EvalDefinition:
    eval_key: str
    display_name: str
    eval_version: str
    result_kind: ResultKind
    labels: tuple[str, ...]
    function: EvalFunction
    condition: ConditionFunction | None
    on_cancel: CancellationFunction | None
    timeout_seconds: float | None

    def catalog_definition(self) -> CatalogDefinition:
        return CatalogDefinition(
            eval_key=self.eval_key,
            display_name=self.display_name,
            eval_version=self.eval_version,
            result_kind=self.result_kind,
            labels=self.labels,
        )


class Evaluator:
    """A process-local collection of explicitly versioned evaluations."""

    def __init__(self, *, name: str, version: str) -> None:
        self.name = _bounded(name, field_name="name", maximum=MAX_DISPLAY_NAME_BYTES)
        self.version = _bounded(
            version, field_name="version", maximum=MAX_VERSION_BYTES
        )
        self._definitions: dict[str, EvalDefinition] = {}

    def eval(
        self,
        eval_key: str,
        *,
        version: str,
        display_name: str | None = None,
        result_kind: ResultKind | str = ResultKind.SCORE,
        labels: tuple[str, ...] | list[str] = (),
        when: ConditionFunction | None = None,
        on_cancel: CancellationFunction | None = None,
        timeout_seconds: float | None = None,
    ) -> Callable[[EvalFunction], EvalFunction]:
        key = _validate_key(eval_key)
        eval_version = _bounded(
            version, field_name="eval version", maximum=MAX_VERSION_BYTES
        )
        display = _bounded(
            display_name or eval_key.replace("_", " ").capitalize(),
            field_name="display name",
            maximum=MAX_DISPLAY_NAME_BYTES,
        )
        kind = ResultKind(result_kind)
        normalized_labels = _labels(list(labels))
        if timeout_seconds is not None:
            timeout_seconds = _finite(timeout_seconds, "timeout_seconds")
            if timeout_seconds <= 0:
                raise ValueError("timeout_seconds must be greater than zero")

        def register(function: EvalFunction) -> EvalFunction:
            if key in self._definitions:
                raise ValueError(f"duplicate eval key: {key}")
            if len(self._definitions) >= MAX_CATALOG_DEFINITIONS:
                raise ValueError(
                    f"an evaluator may define at most {MAX_CATALOG_DEFINITIONS} evaluations"
                )
            if not callable(function):
                raise TypeError("evaluation must be callable")
            if when is not None and not callable(when):
                raise TypeError("when must be callable")
            if on_cancel is not None and not callable(on_cancel):
                raise TypeError("on_cancel must be callable")
            self._definitions[key] = EvalDefinition(
                eval_key=key,
                display_name=display,
                eval_version=eval_version,
                result_kind=kind,
                labels=normalized_labels,
                function=function,
                condition=when,
                on_cancel=on_cancel,
                timeout_seconds=timeout_seconds,
            )
            return function

        return register

    @property
    def definitions(self) -> tuple[EvalDefinition, ...]:
        return tuple(self._definitions[key] for key in sorted(self._definitions))

    def catalog(self) -> tuple[CatalogDefinition, ...]:
        return tuple(definition.catalog_definition() for definition in self.definitions)

    @property
    def catalog_revision(self) -> str:
        payload = [item.to_wire() for item in self.catalog()]
        canonical = json.dumps(
            payload,
            ensure_ascii=False,
            allow_nan=False,
            sort_keys=True,
            separators=(",", ":"),
        ).encode("utf-8")
        return "sha256:" + hashlib.sha256(canonical).hexdigest()

    def definition(self, eval_key: str) -> EvalDefinition:
        try:
            return self._definitions[eval_key]
        except KeyError as error:
            raise KeyError(f"unknown eval key: {eval_key}") from error

    def run_from_env(self) -> None:
        """Run this evaluator until the process receives a stop request."""
        import asyncio
        import signal

        from failproofai_sdk.evaluator.runtime import WorkerConfig, WorkerRuntime

        async def run() -> None:
            runtime = WorkerRuntime(self, WorkerConfig.from_env())
            loop = asyncio.get_running_loop()
            for name in ("SIGINT", "SIGTERM"):
                process_signal = getattr(signal, name, None)
                if process_signal is None:
                    continue
                try:
                    loop.add_signal_handler(process_signal, runtime.stop)
                except (NotImplementedError, RuntimeError):
                    pass
            await runtime.run_forever()

        asyncio.run(run())

    @staticmethod
    async def call(
        function: EvalFunction | ConditionFunction, session: SessionTranscript
    ) -> Any:
        result = function(session)
        if inspect.isawaitable(result):
            return await result
        return result

File: sdk/typescript/src/evaluator/authoring.ts
/**
 * Evaluator definition registry and typed author results.
 *
 * Every bound checked here is a bound the SERVER also checks, and rejects with
 * a NON-RETRYABLE 422. Checking them at authoring time is what turns "a
 * successful evaluation silently dead-lettered in production" into an error at
 * the line that wrote it.
 */

import { createHash } from "node:crypto";

import {
  MAX_CATALOG_DEFINITIONS,
  MAX_DESCRIPTION_BYTES,
  MAX_DISPLAY_NAME_BYTES,
  MAX_DISPLAY_VALUE_BYTES,
  MAX_EVAL_KEY_BYTES,
  MAX_LABEL_BYTES,
  MAX_LABELS_PER_RESULT,
  MAX_REASONING_BYTES,
  MAX_RESULTS_PER_RUN,
  MAX_SUMMARY_BYTES,
  MAX_UNIT_BYTES,
  MAX_VERSION_BYTES,
  ResultKind,
  catalogDefinitionToWire,
} from "./protocol.js";
import type { CatalogDefinition, ResultItem, SessionTranscript } from "./protocol.js";

const KEY_PATTERN = /^[a-z][a-z0-9_]*$/;

export type EvalFunction = (session: SessionTranscript) => EvalResult | Promise<EvalResult>;
export type ConditionFunction = (
  session: SessionTranscript,
) => boolean | ConditionResult | Promise<boolean | ConditionResult>;
export type CancellationFunction = (session: SessionTranscript) => unknown;

/**
 * Compiles server-authored (sandboxed) source into something callable with a
 * session. Signature mirrors `source.compileEvaluator`, which is the default.
 *
 * A host that serves managed definitions whose source is NOT a restricted
 * expression — a declarative judge document, say — installs its own compiler
 * here. The SDK stays agnostic: it never inspects the source, it just hands it
 * to whoever compiles. Returning an async function is supported and is the
 * right shape for a compiler whose work is IO-bound.
 */
export type ManagedCompiler = (
  source: string,
  options: { timeoutSeconds?: number | null; evalKey?: string | null },
) => EvalFunction;

function utf8Length(value: string): number {
  return Buffer.byteLength(value, "utf8");
}

function bounded(value: unknown, fieldName: string, maximum: number): string {
  if (typeof value !== "string") throw new TypeError(`${fieldName} must be a string`);
  if (value === "") throw new Error(`${fieldName} must not be empty`);
  const size = utf8Length(value);
  if (size > maximum) throw new Error(`${fieldName} is ${size} bytes; maximum is ${maximum}`);
  // Reject C0 control characters and DEL, matching the server's own check.
  // Without this the SDK accepts a string — reasoning or a summary quoting
  // transcript text that contains an ANSI escape or a NUL — that the server
  // then rejects with a NON-RETRYABLE 422, so a successful evaluation is
  // silently lost and its assignment dead-letters. TAB, LF and CR are kept
  // because real multi-line reasoning uses them.
  for (const char of value) {
    const code = char.codePointAt(0)!;
    if ((code < 0x20 && char !== "\t" && char !== "\n" && char !== "\r") || code === 0x7f) {
      throw new Error(
        `${fieldName} must not contain control characters (found U+${code
          .toString(16)
          .toUpperCase()
          .padStart(4, "0")})`,
      );
    }
  }
  return value;
}

function finite(value: unknown, fieldName: string): number {
  if (typeof value !== "number") throw new TypeError(`${fieldName} must be a number`);
  if (!Number.isFinite(value)) throw new Error(`${fieldName} must be finite`);
  return value;
}

function normalizeLabels(values: readonly string[]): string[] {
  if (values.length > MAX_LABELS_PER_RESULT) {
    throw new Error(`at most ${MAX_LABELS_PER_RESULT} labels are allowed`);
  }
  const normalized = values.map((label) => bounded(label, "label", MAX_LABEL_BYTES));
  if (new Set(normalized).size !== normalized.length) throw new Error("labels must be unique");
  return [...normalized].sort();
}

function validateResultText(
  unit: string,
  displayValue: string | undefined,
  description: string | undefined,
): void {
  if (unit) bounded(unit, "unit", MAX_UNIT_BYTES);
  if (displayValue !== undefined) bounded(displayValue, "display value", MAX_DISPLAY_VALUE_BYTES);
  if (description !== undefined) bounded(description, "description", MAX_DESCRIPTION_BYTES);
}

export function validateKey(value: string, fieldName = "eval_key"): string {
  bounded(value, fieldName, MAX_EVAL_KEY_BYTES);
  if (!KEY_PATTERN.test(value)) throw new Error(`${fieldName} must match ${KEY_PATTERN.source}`);
  return value;
}

export class Score {
  readonly value: number;
  readonly passed: boolean | undefined;
  readonly unit: string;
  readonly displayValue: string | undefined;
  readonly description: string | undefined;

  constructor(
    value: number,
    options: {
      passed?: boolean;
      unit?: string;
      displayValue?: string;
      description?: string;
    } = {},
  ) {
    const numeric = finite(value, "score value");
    if (numeric < 0 || numeric > 1) throw new Error("score value must be between 0 and 1");
    if (options.passed !== undefined && typeof options.passed !== "boolean") {
      throw new TypeError("score passed must be a boolean or undefined");
    }
    this.value = numeric;
    this.passed = options.passed;
    this.unit = options.unit ?? "ratio";
    this.displayValue = options.displayValue;
    this.description = options.description;
    validateResultText(this.unit, this.displayValue, this.description);
    Object.freeze(this);
  }
}

export class Metric {
  readonly value: number;
  readonly unit: string;
  readonly displayValue: string | undefined;
  readonly description: string | undefined;

  constructor(
    value: number,
    options: { unit?: string; displayValue?: string; description?: string } = {},
  ) {
    this.value = finite(value, "metric value");
    this.unit = options.unit ?? "";
    this.displayValue = options.displayValue;
    this.description = options.description;
    validateResultText(this.unit, this.displayValue, this.description);
    Object.freeze(this);
  }
}

export class Assertion {
  readonly passed: boolean;
  readonly description: string | undefined;

  constructor(passed: boolean, options: { description?: string } = {}) {
    if (typeof passed !== "boolean") throw new TypeError("assertion passed must be a boolean");
    this.passed = passed;
    this.description = options.description;
    if (this.description !== undefined) {
      bounded(this.description, "description", MAX_DESCRIPTION_BYTES);
    }
    Object.freeze(this);
  }
}

export class ConditionResult {
  readonly applicable: boolean;
  readonly reasonCode: string;

  constructor(applicable: boolean, reasonCode = "condition_false") {
    if (typeof applicable !== "boolean") {
      throw new TypeError("condition applicable must be a boolean");
    }
    this.applicable = applicable;
    this.reasonCode = reasonCode;
    validateKey(reasonCode, "condition reason code");
    Object.freeze(this);
  }
}

export interface EvalResultOptions {
  score?: Score;
  metrics?: Record<string, Metric | number>;
  assertions?: Record<string, Assertion | boolean>;
  reasoning?: string;
  summary?: string;
  labels?: readonly string[];
}

export class EvalResult {
  readonly score: Score | undefined;
  readonly metrics: Readonly<Record<string, Metric | number>>;
  readonly assertions: Readonly<Record<string, Assertion | boolean>>;
  readonly reasoning: string | undefined;
  readonly summary: string | undefined;
  readonly labels: readonly string[];

  constructor(options: EvalResultOptions = {}) {
    this.score = options.score;
    this.metrics = options.metrics ?? {};
    this.assertions = options.assertions ?? {};
    this.reasoning = options.reasoning;
    this.summary = options.summary;
    if (this.reasoning !== undefined) bounded(this.reasoning, "reasoning", MAX_REASONING_BYTES);
    if (this.summary !== undefined) bounded(this.summary, "summary", MAX_SUMMARY_BYTES);
    this.labels = normalizeLabels(options.labels ?? []);
    Object.freeze(this);
  }

  resultItems(evalKey: string): ResultItem[] {
    const items: ResultItem[] = [];
    if (this.score !== undefined) {
      items.push({
        resultKey: evalKey,
        resultKind: ResultKind.SCORE,
        numericValue: this.score.value,
        boolValue: this.score.passed ?? null,
        unit: this.score.unit,
        displayValue: this.score.displayValue ?? null,
        description: this.score.description ?? null,
        reasoning: this.reasoning ?? null,
        labels: this.labels,
      });
    }
    for (const key of Object.keys(this.metrics).sort()) {
      validateKey(key, "metric key");
      const raw = this.metrics[key]!;
      const metric = raw instanceof Metric ? raw : new Metric(raw);
      items.push({
        resultKey: key,
        resultKind: ResultKind.METRIC,
        numericValue: metric.value,
        unit: metric.unit,
        displayValue: metric.displayValue ?? null,
        description: metric.description ?? null,
        // A metric-kind eval's primary result IS the metric whose key equals
        // `evalKey`; attach the eval's reasoning there so it is not silently
        // dropped for non-score evals.
        reasoning: key === evalKey ? (this.reasoning ?? null) : null,
        labels: this.labels,
      });
    }
    for (const key of Object.keys(this.assertions).sort()) {
      validateKey(key, "assertion key");
      const raw = this.assertions[key]!;
      const assertion = raw instanceof Assertion ? raw : new Assertion(raw);
      items.push({
        resultKey: key,
        resultKind: ResultKind.ASSERTION,
        boolValue: assertion.passed,
        description: assertion.description ?? null,
        reasoning: key === evalKey ? (this.reasoning ?? null) : null,
        labels: this.labels,
      });
    }
    if (items.length === 0) {
      throw new Error("an EvalResult must contain a score, metric, or assertion");
    }
    if (items.length > MAX_RESULTS_PER_RUN) {
      throw new Error(`an EvalResult may contain at most ${MAX_RESULTS_PER_RUN} results`);
    }
    const keys = items.map((item) => item.resultKey);
    if (new Set(keys).size !== keys.length) {
      throw new Error("result keys must be unique within one evaluation run");
    }
    return items;
  }
}

export interface EvalDefinition {
  evalKey: string;
  displayName: string;
  evalVersion: string;
  resultKind: ResultKind;
  labels: readonly string[];
  function: EvalFunction;
  condition: ConditionFunction | null;
  onCancel: CancellationFunction | null;
  timeoutSeconds: number | null;
}

export function catalogDefinitionOf(definition: EvalDefinition): CatalogDefinition {
  return {
    evalKey: definition.evalKey,
    displayName: definition.displayName,
    evalVersion: definition.evalVersion,
    resultKind: definition.resultKind,
    labels: definition.labels,
  };
}

/** Deterministic JSON with sorted keys and no spaces, for the catalog hash. */
function canonicalJson(value: unknown): string {
  if (value === null || typeof value !== "object") return JSON.stringify(value) ?? "null";
  if (Array.isArray(value)) return `[${value.map(canonicalJson).join(",")}]`;
  const entries = Object.entries(value as Record<string, unknown>)
    .filter(([, item]) => item !== undefined)
    .sort(([a], [b]) => (a < b ? -1 : a > b ? 1 : 0));
  return `{${entries.map(([key, item]) => `${JSON.stringify(key)}:${canonicalJson(item)}`).join(",")}}`;
}

export interface EvalOptions {
  version: string;
  displayName?: string;
  resultKind?: ResultKind;
  labels?: readonly string[];
  when?: ConditionFunction;
  onCancel?: CancellationFunction;
  timeoutSeconds?: number;
}

/**
 * Marks an `Evaluator` across the package's two builds.
 *
 * `failproofai-evaluator` is the ESM build, and a CommonJS evals file (plain
 * `tsc` output, `require`) constructs its `Evaluator` from `dist/cjs` — a
 * second, unrelated copy of the class. `instanceof` is false across the two, so
 * the loader refused the commonest setup there is with "resolved to Evaluator,
 * not an Evaluator". `Symbol.for` is one registry per process, so both copies
 * stamp and read the same key.
 */
const EVALUATOR_BRAND = Symbol.for("@failproofai/sdk/evaluator.Evaluator");

/**
 * True for an `Evaluator` from either build of this package.
 *
 * Safe to hand the result to `runFromEnv()`: that method belongs to the copy
 * that built the object, and runs the runtime from that same copy, so the
 * result classes it checks are the ones the evaluations construct.
 *
 * @internal
 */
export function isEvaluator(value: unknown): value is Evaluator {
  return (
    typeof value === "object" &&
    value !== null &&
    (value as Record<symbol, unknown>)[EVALUATOR_BRAND] === true &&
    typeof (value as { runFromEnv?: unknown }).runFromEnv === "function"
  );
}

/** A process-local collection of explicitly versioned evaluations. */
export class Evaluator {
  declare readonly [EVALUATOR_BRAND]: true;
  readonly name: string;
  readonly version: string;
  readonly managedCompiler: ManagedCompiler | null;
  private readonly registry = new Map<string, EvalDefinition>();

  constructor(options: { name: string; version: string; managedCompiler?: ManagedCompiler }) {
    Object.defineProperty(this, EVALUATOR_BRAND, { value: true });
    this.name = bounded(options.name, "name", MAX_DISPLAY_NAME_BYTES);
    this.version = bounded(options.version, "version", MAX_VERSION_BYTES);
    // Omitted keeps the default: server-authored source is compiled and
    // sandboxed by `source.compileEvaluator`. Customer workers never set this —
    // their definitions are LOCAL and never carry source at all.
    if (options.managedCompiler !== undefined && typeof options.managedCompiler !== "function") {
      throw new TypeError("managedCompiler must be a function");
    }
    this.managedCompiler = options.managedCompiler ?? null;
  }

  /**
   * Register one evaluation.
   *
   *     evaluator.eval("tool_success_rate", { version: "1" }, (session) =>
   *       new EvalResult({ score: new Score(ratio) }),
   *     );
   */
  eval(evalKey: string, options: EvalOptions, fn: EvalFunction): EvalFunction {
    const key = validateKey(evalKey);
    const evalVersion = bounded(options.version, "eval version", MAX_VERSION_BYTES);
    const display = bounded(
      options.displayName ?? defaultDisplayName(evalKey),
      "display name",
      MAX_DISPLAY_NAME_BYTES,
    );
    const kind = options.resultKind ?? ResultKind.SCORE;
    if (!Object.values(ResultKind).includes(kind)) {
      throw new Error(`resultKind must be one of ${Object.values(ResultKind).join(", ")}`);
    }
    const labels = normalizeLabels(options.labels ?? []);
    let timeoutSeconds: number | null = null;
    if (options.timeoutSeconds !== undefined) {
      timeoutSeconds = finite(options.timeoutSeconds, "timeoutSeconds");
      if (timeoutSeconds <= 0) throw new Error("timeoutSeconds must be greater than zero");
    }

    if (this.registry.has(key)) throw new Error(`duplicate eval key: ${key}`);
    if (this.registry.size >= MAX_CATALOG_DEFINITIONS) {
      throw new Error(
        `an evaluator may define at most ${MAX_CATALOG_DEFINITIONS} evaluations`,
      );
    }
    if (typeof fn !== "function") throw new TypeError("evaluation must be a function");
    if (options.when !== undefined && typeof options.when !== "function") {
      throw new TypeError("when must be a function");
    }
    if (options.onCancel !== undefined && typeof options.onCancel !== "function") {
      throw new TypeError("onCancel must be a function");
    }

    this.registry.set(key, {
      evalKey: key,
      displayName: display,
      evalVersion,
      resultKind: kind,
      labels,
      function: fn,
      condition: options.when ?? null,
      onCancel: options.onCancel ?? null,
      timeoutSeconds,
    });
    return fn;
  }

  get definitions(): readonly EvalDefinition[] {
    return [...this.registry.keys()].sort().map((key) => this.registry.get(key)!);
  }

  catalog(): CatalogDefinition[] {
    return this.definitions.map(catalogDefinitionOf);
  }

  get catalogRevision(): string {
    const payload = this.catalog().map(catalogDefinitionToWire);
    const canonical = Buffer.from(canonicalJson(payload), "utf8");
    return `sha256:${createHash("sha256").update(canonical).digest("hex")}`;
  }

  definition(evalKey: string): EvalDefinition {
    const found = this.registry.get(evalKey);
    if (found === undefined) throw new Error(`unknown eval key: ${evalKey}`);
    return found;
  }

  /** Run this evaluator until the process receives a stop request. */
  async runFromEnv(): Promise<void> {
    const { WorkerRuntime, workerConfigFromEnv } = await import("./runtime.js");
    const runtime = new WorkerRuntime(this, workerConfigFromEnv());
    const stop = (): void => {
      runtime.stop();
    };
    for (const signal of ["SIGINT", "SIGTERM"] as const) {
      process.once(signal, stop);
    }
    try {
      await runtime.runForever();
    } finally {
      for (const signal of ["SIGINT", "SIGTERM"] as const) {
        process.removeListener(signal, stop);
      }
    }
  }

  /** Call an evaluation or condition, awaiting it when it returns a promise. */
  static async call(
    fn: EvalFunction | ConditionFunction,
    session: SessionTranscript,
  ): Promise<unknown> {
    return await (fn as (s: SessionTranscript) => unknown)(session);
  }
}

function defaultDisplayName(evalKey: string): string {
  const spaced = evalKey.replaceAll("_", " ");
  return spaced.charAt(0).toUpperCase() + spaced.slice(1);
}
